// fountain-js/dist/lexer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InlineLexer = exports.Lexer = void 0;
const rules_1 = require("./rules");
const utilities_1 = require("./utilities");
class Lexer {
    reconstruct(script) {
        return script.replace(rules_1.rules.boneyard, '\n$1\n')
            .replace(rules_1.rules.standardizer, '\n')
            .replace(rules_1.rules.cleaner, '')
            .replace(rules_1.rules.whitespacer, '');
    }
}
exports.Lexer = Lexer;
class InlineLexer extends Lexer {
    constructor() {
        super(...arguments);
        this.inline = {
            note: '<!-- $1 -->',
            line_break: '<br />',
            bold_italic_underline: '<span class="bold italic underline">$1</span>',
            bold_underline: '<span class="bold underline">$1</span>',
            italic_underline: '<span class="italic underline">$1</span>',
            bold_italic: '<span class="bold italic">$1</span>',
            bold: '<span class="bold">$1</span>',
            italic: '<span class="italic">$1</span>',
            underline: '<span class="underline">$1</span>',
            escape: '$1'
        };
    }
    reconstruct(line) {
        if (!line)
            return;
        let match;
        const styles = ['bold_italic_underline', 'bold_underline', 'italic_underline', 'bold_italic', 'bold', 'italic', 'underline'];
        line = utilities_1.escapeHTML(line
            .replace(rules_1.rules.note_inline, this.inline.note)
            .replace(rules_1.rules.escape, '[{{{$&}}}]') // perserve escaped characters
        );
        for (let style of styles) {
            match = rules_1.rules[style];
            if (match.test(line)) {
                line = line.replace(match, this.inline[style]);
            }
        }
        return line
            .replace(/\n/g, this.inline.line_break)
            .replace(/\[{{{\\(&.+?;|.)}}}]/g, this.inline.escape) // restore escaped chars to intended sequence
            .trim();
    }
}
exports.InlineLexer = InlineLexer;
